invenio_users_resources/records/api.py

```diff
@@ -0,0 +1,234 @@
+    data = {
+        "id": user.id,
+        "email": user.email,
+        "username": None,
+        "full_name": None,
+    }
+
+    if user.profile is not None:
+        data["full_name"] = user.profile.full_name
+        data["username"] = user.profile.username
+        # TODO populate more when we have extensible user profiles
+
+    return data
+
+
+def parse_role_data(role):
+    """Parse the role's information into a dictionary."""
+    data = {"id": role.id, "name": role.name, "description": role.description}
+    return data
+
+
+class UserAggregate(Record):
+    """An aggregate of information about a user."""
+
+    model_cls = UserAggregateModel
+    """The model class for the request."""
+
+    # NOTE: the "uuid" isn't a UUID but contains the same value as the "id"
+    #       field, which is currently an integer for User objects!
+    dumper = ElasticsearchDumper(
+        extensions=[], model_fields={"id": ("uuid", int)}
+    )
+    """Elasticsearch dumper with configured extensions."""
+
+    metadata = None
+    """Disabled metadata field from the base class."""
+
+    schema = ConstantField("$schema", "local://users/user-v1.0.0.json")
+    """The JSON Schema to use for validation."""
+
+    index = IndexField("users-user-v1.0.0", search_alias="users")
+    """The Elasticsearch index to use."""
+
+    # TODO
+    id = ModelField("id")
+    """The data-layer id."""
+
+    email = DictField("email")
+    """The user's email address."""
+
+    username = DictField("username")
+    """The user's username."""
+
+    full_name = DictField("full_name")
+    """The user's full name."""
+
+    _user = None
+    """The cached User entity."""
+
+    @property
+    def user(self):
+        """Cache for the associated user object."""
+        user = self._user
+        if user is None:
+            id_, email = self.id, self.email
+            user = current_datastore.get_user(id_ or email)
+
+            self._user = user
+
+        return user
+
+    @classmethod
+    def create(
+        cls, data, id_=None, validator=None, format_checker=None, **kwargs
+    ):
+        # NOTE: we don't use an actual database table, and as such can't
+        #       use db.session.add(record.model)
+        # TODO: should we do some JSON Schema validation here?
+        #       or do we rely on marshmallow?
+        with db.session.begin_nested():
+            email = data["email"]
+            active = data.pop("active", True)
+
+            full_name = data.pop("full_name", None)
+            username = data.pop("username", None)
+
+            # TODO populate user profile
+            if full_name and username:
+                pass
+
+            user = current_datastore.create_user(email=email, active=active)
+            user_aggregate = cls.from_user(user)
+
+            return user_aggregate
+
+    def _validate(self, *args, **kwargs):
+        """Skip the validation."""
+        # TODO does a JSONSchema even make sense? we don't actually store
```

Comment:
Indeed, could be we simply reply on the SQLAlchemy models  which defines the database table structure.